File: src/main/groovy/com/github/camork/filesystem/zip/ZipFileSystem.java
// Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.github.camork.filesystem.zip;

import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.VirtualFileManager;
import com.intellij.openapi.vfs.impl.jar.JarFileSystemImpl;

import org.jetbrains.annotations.NotNull;

public abstract class ZipFileSystem extends JarFileSystemImpl {

    public static final String protocol = "zip";

    public static ZipFileSystem getInstance() {
        return (ZipFileSystem)VirtualFileManager.getInstance().getFileSystem(protocol);
    }

    @Override
    protected boolean isCorrectFileType(@NotNull VirtualFile local) {
        return true;
    }

}

File: src/main/kotlin/com/github/camork/filesystem/zip/ZipFileSystem.kt
package com.github.camork.filesystem.zip

import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.vfs.VirtualFileManager
import com.intellij.openapi.vfs.impl.jar.JarFileSystemImpl

/**
 * @author Charles Wu
 */
abstract class ZipFileSystem : JarFileSystemImpl() {

    companion object {
        const val PROTOCOL = "zip"

        fun getInstance(): ZipFileSystem {
            return VirtualFileManager.getInstance().getFileSystem(PROTOCOL) as ZipFileSystem
        }
    }

    override fun isCorrectFileType(local: VirtualFile): Boolean {
        return true
    }
}

class ZipFileSystemImpl : ZipFileSystem() {

    override fun getProtocol(): String = PROTOCOL
}